### PR TITLE
Issue/350 change load python

### LINKED ISSUE #350: Make METdatadb the primary loading method
## Describe the Task ##
Currently METviewer Java classes are used as a default to load data to the database.
Need to change the mv_load.sh to use METdatadb instead, remve mv_load_pythn.sh and create mv_load_java.sh
Also need to modify METviewer regression testing procedure by replacing Java loader with Python 

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 2 days of work.*

### Sub-Issues ###
Consider breaking the task down into sub-issues.
- [x] Change scripts
- [x] Modify testing
- [x] Modify Dockerfile

### Relevant Deadlines ###
*NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Task Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by TatianaBurek**
## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>
- tested Docker
- tested regression testing script

- [X] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
Steps to test Docker container
- test Docker file and make sure that the image is build and the container is started 
- verify that METdatadb is installed in /METviewer-python
- verify that /METviewer/bin/mv_load.sh uses Python and PYTHON_ENV and METDATADB_HOME are set correctly
- try to run  mv_load.sh 

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**

- [X] Do these changes include sufficient testing updates? **[No]**

- [X] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Review by JohnHalleyGotway — CHANGES_REQUESTED** [review]
@TatianaBurek I think you need to update the "hooks/build" file to add:
```
--build-arg METDATADB_GIT_NAME=v1.0.0
```
Otherwise the DockerHub build will fail.


After adding it to my test build, it seems to be working (at least it's in the process of building the image).
```
cd METviewer/docker
git checkout issue/350-change_load_python
docker build -t metviewer:feature_350 \
--build-arg METVIEWER_GIT_NAME=issue/350-change_load_python \
--build-arg METCALCPY_GIT_NAME=v1.0.0 \
--build-arg METPLOTPY_GIT_NAME=v1.0.0 \
--build-arg METDATADB_GIT_NAME=develop .
```
As @TatianaBurek advises, I'm using "develop" for testing.

File: bin/auto_test.sh [hunk 1]
```diff
@@ -43,13 +43,15 @@ export git_user=""
...
 export PYTHON_ENV=""
 export METCALCPY_HOME=""
 export METPLOTPY_HOME=""
+export METDATADB_HOME=""
 
-while getopts "U:t:b:B:l:d:m:a:g:G:u:p:h:P:s:j:c:T:E:A:R?" o; do
+while getopts "U:t:b:B:l:d:m:a:g:G:u:p:h:P:s:j:c:T:E:A:R:D:L:?" o; do
     case "${o}" in
-    	c)
+    	  c)
```
Comment (JohnHalleyGotway) [inline bin/auto_test.sh]:
```suggestion
        c)
```
Just consistent spacing... although this is a drop in the bucket. Looks like there a big mix of whitespace types below.

**Comment by JohnHalleyGotway** [comment]
After reviewing these changes, I added @venitahagerty to the list of reviewers as well. @venitahagerty please just review the code changes so that you're aware of them... fully switching over to redirect mv_load.sh to call the METdatadb loader instead.

**Comment by TatianaBurek** [comment]
I added METDATADB_GIT_NAME to "hooks/build" file

**Review by JohnHalleyGotway — COMMENTED** [review]
Thanks for updating the build hook @TatianaBurek. I reviewed the code changes and see...
- removal of several test utilities related to loading
- updates for newer java versions, esp with log4j
- switching mv_load.sh to redirect to calling the METdatadb loader

I tested this by building a Docker image for this feature branch locally.

Then I setup a test env.
```
METVIEWER_DIR=/Volumes/d1/projects/METviewer/docker/test_mv/metv
MYSQL_DIR=/Volumes/d1/projects/METviewer/docker/test_mv/mysql
METVIEWER_DOCKER_IMAGE=metviewer:feature_350
METVIEWER_DATA=/Volumes/d1/projects/METviewer/docker/test_mv/data
```

Then I brought it up.
```
docker-compose up -d
```

And confirmed that the GUI is accessible at: http://localhost:8080/metviewer/metviewer1.jsp

I referred to these [training instructions](https://metplus-training.readthedocs.io/en/latest/modules/METviewer/docker.html) and ran.
```
cd $METVIEWER_DATA
curl -SL https://dtcenter.ucar.edu/dfiles/code/METplus/METviewer/sample_data-met_out_v9.1.tgz | tar -xzC .
docker exec -it metviewer_1 /bin/bash
mysql -hmysql_mv -uroot -pmvuser -e"create database mv_met_out;"
mysql -hmysql_mv -uroot -pmvuser mv_met_out < /METviewer/sql/mv_mysql.sql
/METviewer/bin/mv_load.sh /data/load_met_out.xml
```

So this definitely calls METdatadb:
```
INFO:root:METdbload Version: 1.0.0
INFO:root:--- *** --- Start METdbLoad --- *** ---
INFO:root:Begin time: 2022-01-12 20:59:13.502443
WARNING:root:!!! Unknown tag: met_version
```

However it does error out:
```
ValueError: 'fcst_valid' is not in list
```

I suspect that this branch in METviewer is doing what it is supposed to do. However, this error is either the result of a problem in METdatadb or a problem in my XML load spec.

Any advice?

**Comment by venitahagerty** [comment]
@JohnHalleyGotway It would be helpful for me to see the whole log output. It sounds like you might have fcst_valid in a table where it is not expected. I might need the xml and the input data as well.

**Comment by TatianaBurek** [comment]
Could it be that we use METdatadb 1.0.0 and the DB schema from METviewer 4.1.0-beta?
For this test we should use METdatadb 'develop'

**Comment by JohnHalleyGotway** [comment]
@TatianaBurek I rebuilt METviewer using METdatadb:develop, but got the same result.

@venitahagerty this data is referenced in this [training video](https://metplus-training.readthedocs.io/en/latest/modules/METviewer/docker.html).

In that video, we have users pull/untar this data:
   https://dtcenter.ucar.edu/dfiles/code/METplus/METviewer/sample_data-met_out_v9.1.tgz

And that tar file includes an XML load spec named "load_met_out.xml"
[load_met_out.xml.txt](https://github.com/dtcenter/METviewer/files/7858811/load_met_out.xml.txt). I commented out <met_version> because the loader printed a warning about it.

And here's the full log I'm getting when running mv_load.sh inside the container:
[mv_load.log](https://github.com/dtcenter/METviewer/files/7858815/mv_load.log)

This is MET version 9.1 output and we load it inside the container. If you load it outside of a container, you'll need to modify the <folder_tmpl> in the XML load spec.

The question are...
Should this older load spec still work or does something need to be updated?
If it should still work, is METdatadb making some assumption about the way the data is organized? Or does is it having problems loading MET version 9.1 data?

**Comment by venitahagerty** [comment]
Note that the tag verbose can be set to true and you'll get more output.

From my work so far, it looks like METdbload thinks it has a mode file - which is weird. I think that's in reference to grid_stat_POP_12_1080000L_20050808_000000V.stat. I'll continue looking into this.

**Comment by venitahagerty** [comment]
@JohnHalleyGotway I'm seeing files like grid_stat_120000L_20050807_120000V_cts.txt in met_out/grid_stat. I also see a couple of lines in read_data_files.py: 

elif lc_filename.endswith("cts.txt"):
            lu_type = CN.MODE_CTS

So it's identifying an individual stat file as a mode file. I'm going to review file names to see what the differences are supposed to be. Would you want the individual line type files also loaded - though it seems like they would be duplicates? Ignored?

**Comment by JohnHalleyGotway** [comment]
Venita,

We should be loading:

- All *.stat files which are created by point_stat, grid_stat,
wavelet_stat, and so on.
- All mode*.txt files created by mode.

The optional grid_stat*.txt files should be ignored.

We could consider meeting with Tatiana to review the existing load logic
for METviewer… and then write down the details in the METdatadb
documentation. If that’d be useful please feel free to schedule it or let
me know and I can.

Thanks
John

**Comment by venitahagerty** [comment]
Added an additional check to read_data_file.py in METdatadb/METdbLoad/ush for cts mode files. That seems to have solved the problem of mistaking optional grid_stat*.txt files for Mode files, but have revealed some additional data processing problems.

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes.

While my testing was not entirely successful, I am satisfied that these changes to METviewer are doing what was stated in issue dtcenter/METviewer#350. Recommend merging this PR now and proceeding with the beta5 development release.

@venitahagerty I see that you've identified some additional work that's required for in dtcenter/METdatadb#68. Once that issue and any other related issues that arise in your testing, we'll need to redo this METdatadb/METviewer integration testing to confirm that they're working well together.

**Pull request merged**